health-check failed on server with defined <http_server_default_response> option

Many months ago, as defined in [clickhouse docs](https://clickhouse.yandex/docs/en/operations/server_settings/settings/#http-server-default-response) i add variable http_server_default_response to clickhouse config for starting tabix in browser. Now i need to setup chproxy and see in log for every clickhouse node this:

```
ERROR: 2019/05/16 07:59:00 scope.go:553: error while health-checking "10.60.11.241:8123" host: unexpected response: <html ng-app="SMI2"><head><base href="http://ui.tabix.io/"></head><body><div ui-view="" class="content-ui"></div><script src="http://loader.tabix.io/master.js"></script></body></html>
```

my config.yml:

```
hack_me_please: true
log_debug: true

server:
  http:
      listen_addr: ":9191"

users:
  - name: "grafana"
    password: "***"
    to_cluster: "click-na"
    to_user: "grafana"
    max_concurrent_queries: 20
    max_execution_time: 120s
    requests_per_minute: 50
    allow_cors: true
    max_queue_size: 40
    max_queue_time: 30s
    cache: "shortterm"

clusters:
  - name: "click-na"
    nodes: [
     "10.60.11.241:8123",
     "10.60.11.242:8123",
     "10.60.11.243:8123",
    ]
    users:
      - name: "grafana"
        password: "***"

caches:
  - name: "shortterm"
    dir: "/tmp/cache"
    max_size: 150Mb
    expire: 600s
```

curl request to / of clickhouse node:

```
# curl -v 10.60.11.241:8123/
*   Trying 10.60.11.241...
* TCP_NODELAY set
* Connected to 10.60.11.241 (10.60.11.241) port 8123 (#0)
> GET / HTTP/1.1
> Host: 10.60.11.241:8123
> User-Agent: curl/7.52.1
> Accept: */*
> 
< HTTP/1.1 200 OK
< Date: Thu, 16 May 2019 08:20:29 GMT
< Connection: Keep-Alive
< Keep-Alive: timeout=3
< Content-Type: text/html; charset=UTF-8
< Content-Length: 183
< 
* Curl_http_done: called premature == 0
* Connection #0 to host 10.60.11.241 left intact
<html ng-app="SMI2"><head><base href="http://ui.tabix.io/"></head><body><div ui-view="" class="content-ui"></div><script src="http://loader.tabix.io/master.js"></script></body></html>
```

I think, simple request to '/' and search 'Ok' in body is not good for database healthcheck.
May be `SELECT 1;` request will be more suitable, because it always return predefined result.

For example:

```
# echo "SELECT 1;" | curl -u grafana:*** --data-binary @- 10.60.11.241:8123
1
```

--- a/README.md
+++ b/README.md
@@ -560,9 +560,31 @@ clusters:
     # Requests are evenly distributed among them.
     nodes: ["127.0.0.1:8123", "shard2:8123"]
 
-    # Each cluster node is checked for availability using this interval.
+    # DEPRECATED: Each cluster node is checked for availability using this interval.
     # By default each node is checked for every 5 seconds.
+    # Use `heartbeat.interval`.
     heartbeat_interval: 1m
+
+    # User configuration for heart beat requests.
+    heartbeat:
+      # An interval for checking all cluster nodes for availability
+      # By default each node is checked for every 5 seconds.
+      interval: 1m
+
+      # A timeout of wait response from cluster nodes
+      # By default 3s
+      timeout: 10s
+
+      # The parameter to set the URI to request in a health check
+      # By default "/ping"
+      request: "/?query=SELECT%201"
+
+      # Reference response from clickhouse on health check request
+      # By default "Ok.\n"
+      response: "1\n"
+
+      # The name of cluster_user whom credentials will be used for heart beat request to clickhouse.
+      to_user: "web"
 
     # Timed out queries are killed using this user.
     # By default `default` user is used.

--- a/README.md
+++ b/README.md
@@ -566,6 +566,7 @@ clusters:
     heartbeat_interval: 1m
 
     # User configuration for heart beat requests.
+    # Credentials of the first user in clusters.users will be used for heart beat requests to clickhouse.
     heartbeat:
       # An interval for checking all cluster nodes for availability
       # By default each node is checked for every 5 seconds.
@@ -576,15 +577,12 @@ clusters:
       timeout: 10s
 
       # The parameter to set the URI to request in a health check
-      # By default "/ping"
-      request: "/?query=SELECT%201"
+      # By default "/?query=SELECT%201"
+      request: "/?query=SELECT%201%2B1"
 
       # Reference response from clickhouse on health check request
-      # By default "Ok.\n"
-      response: "1\n"
-
-      # The name of cluster_user whom credentials will be used for heart beat request to clickhouse.
-      to_user: "web"
+      # By default "1\n"
+      response: "2\n"
 
     # Timed out queries are killed using this user.
     # By default `default` user is used.

--- a/config/README.md
+++ b/config/README.md
@@ -253,8 +253,12 @@ users:
 # By default timed out queries are killed from `default` user.
 kill_query_user: <kill_query_user_config> | optional
 
-# An interval for checking all cluster nodes for availability
+# DEPRECATED: An interval for checking all cluster nodes for availability
+# Use `heartbeat.interval`.
 heartbeat_interval: <duration> | optional | default = 5s
+
+# HeartBeat - user configuration for heart beat requests.
+heartbeat: <heartbeat_config> | optional
 ```
 
 ### <replica_config>
@@ -303,4 +307,22 @@ name: <string>
 
 # User password to access CH with basic auth
 password: <string> | optional
+```
+
+### <heartbeat_config>
+```yml
+# An interval for checking all cluster nodes for availability
+interval: <duration> | optional | default = 5s
+
+# A timeout of wait response from cluster nodes
+timeout: <duration> | optional | default = 3s
+
+# The parameter to set the URI to request in a health check
+request: <string> | optional | default = `/ping`
+
+# Reference response from clickhouse on health check request
+response: <string> | optional | default = `Ok.\n`
+
+# The name of cluster_user whom credentials will be used for heart beat request to clickhouse.
+to_user: <string> | optional
 ```

--- a/config/README.md
+++ b/config/README.md
@@ -318,11 +318,8 @@ interval: <duration> | optional | default = 5s
 timeout: <duration> | optional | default = 3s
 
 # The parameter to set the URI to request in a health check
-request: <string> | optional | default = `/ping`
+request: <string> | optional | default = `/?query=SELECT%201`
 
 # Reference response from clickhouse on health check request
-response: <string> | optional | default = `Ok.\n`
-
-# The name of cluster_user whom credentials will be used for heart beat request to clickhouse.
-to_user: <string> | optional
+response: <string> | optional | default = `1\n`
 ```

--- a/config/config.go
+++ b/config/config.go
@@ -26,8 +26,8 @@ var (
 	defaultHeartBeat = HeartBeat{
 		Interval: Duration(time.Second * 5),
 		Timeout: Duration(time.Second * 3),
-		Request: "/ping",
-		Response: "Ok.\n",
+		Request: "/?query=SELECT%201",
+		Response: "1\n",
 	}
 )
 
@@ -419,10 +419,6 @@ type HeartBeat struct {
     // Reference response from clickhouse on health check request
     // default value is `Ok.\n`
     Response string `yaml:"response,omitempty"`
-
-    // ToUser is the name of cluster_user
-    // whom credentials will be used for heart beat request to clickhouse
-    ToUser string `yaml:"to_user,omitempty"`
 
     // Catches all undefined fields
     XXX map[string]interface{} `yaml:",inline"`

--- a/config/config.go
+++ b/config/config.go
@@ -25,8 +25,8 @@ var (
 
 	defaultHeartBeat = HeartBeat{
 		Interval: Duration(time.Second * 5),
-		Timeout: Duration(time.Second * 3),
-		Request: "/?query=SELECT%201",
+		Timeout:  Duration(time.Second * 3),
+		Request:  "/?query=SELECT%201",
 		Response: "1\n",
 	}
 )
@@ -343,10 +343,13 @@ func (c *Cluster) UnmarshalYAML(unmarshal func(interface{}) error) error {
 		return fmt.Errorf("`cluster.scheme` must be `http` or `https`, got %q instead for %q", c.Scheme, c.Name)
 	}
 	if c.HeartBeatInterval != 0 && c.HeartBeat.Interval != defaultHeartBeat.Interval {
-		return fmt.Errorf("cannot be use `heartbeat_interval` with `heartbeat.interval`")
+		return fmt.Errorf("cannot be use `heartbeat_interval` with `heartbeat` section")
 	}
 	if c.HeartBeat.Interval == 0 && c.HeartBeat.Timeout == 0 && c.HeartBeat.Response == "" {
 		return fmt.Errorf("`cluster.heartbeat` cannot be unset for %q", c.Name)
+	}
+	if c.HeartBeatInterval != 0 && c.HeartBeat.Interval == defaultHeartBeat.Interval {
+		c.HeartBeat.Interval = c.HeartBeatInterval
 	}
 	return checkOverflow(c.XXX, fmt.Sprintf("cluster %q", c.Name))
 }
@@ -403,34 +406,34 @@ func (u *KillQueryUser) UnmarshalYAML(unmarshal func(interface{}) error) error {
 }
 
 type HeartBeat struct {
-    // Interval is an interval of checking
-    // all cluster nodes for availability
-    // if omitted or zero - interval will be set to 5s
-    Interval Duration `yaml:"interval,omitempty"`
+	// Interval is an interval of checking
+	// all cluster nodes for availability
+	// if omitted or zero - interval will be set to 5s
+	Interval Duration `yaml:"interval,omitempty"`
 
-    // Timeout is a timeout of wait response from cluster nodes
-    // if omitted or zero - interval will be set to 3s
-    Timeout Duration `yaml:"timeout,omitempty"`
+	// Timeout is a timeout of wait response from cluster nodes
+	// if omitted or zero - interval will be set to 3s
+	Timeout Duration `yaml:"timeout,omitempty"`
 
-    // Request is a query
-    // default value is `/ping`
-    Request string `yaml:"request,omitempty"`
+	// Request is a query
+	// default value is `/ping`
+	Request string `yaml:"request,omitempty"`
 
-    // Reference response from clickhouse on health check request
-    // default value is `Ok.\n`
-    Response string `yaml:"response,omitempty"`
+	// Reference response from clickhouse on health check request
+	// default value is `Ok.\n`
+	Response string `yaml:"response,omitempty"`
 
-    // Catches all undefined fields
-    XXX map[string]interface{} `yaml:",inline"`
+	// Catches all undefined fields
+	XXX map[string]interface{} `yaml:",inline"`
 }
 
 // UnmarshalYAML implements the yaml.Unmarshaler interface.
 func (h *HeartBeat) UnmarshalYAML(unmarshal func(interface{}) error) error {
-    type plain HeartBeat
-    if err := unmarshal((*plain)(h)); err != nil {
-        return err
-    }
-    return checkOverflow(h.XXX, "heartbeat")
+	type plain HeartBeat
+	if err := unmarshal((*plain)(h)); err != nil {
+		return err
+	}
+	return checkOverflow(h.XXX, "heartbeat")
 }
 
 // User describes list of allowed users

--- a/config/config.go
+++ b/config/config.go
@@ -16,10 +16,18 @@ var (
 	defaultCluster = Cluster{
 		Scheme:       "http",
 		ClusterUsers: []ClusterUser{defaultClusterUser},
+		HeartBeat:    defaultHeartBeat,
 	}
 
 	defaultClusterUser = ClusterUser{
 		Name: "default",
+	}
+
+	defaultHeartBeat = HeartBeat{
+		Interval: Duration(time.Second * 5),
+		Timeout: Duration(time.Second * 3),
+		Request: "/ping",
+		Response: "Ok.\n",
 	}
 )
 
@@ -300,10 +308,13 @@ type Cluster struct {
 	// By default timed out queries are killed under `default` user.
 	KillQueryUser KillQueryUser `yaml:"kill_query_user,omitempty"`
 
-	// HeartBeatInterval is an interval of checking
+	// DEPRECATED: HeartBeatInterval is an interval of checking
 	// all cluster nodes for availability
 	// if omitted or zero - interval will be set to 5s
 	HeartBeatInterval Duration `yaml:"heartbeat_interval,omitempty"`
+
+	// HeartBeat - user configuration for heart beat requests
+	HeartBeat HeartBeat `yaml:"heartbeat,omitempty"`
 
 	// Catches all undefined fields
 	XXX map[string]interface{} `yaml:",inline"`
@@ -331,8 +342,11 @@ func (c *Cluster) UnmarshalYAML(unmarshal func(interface{}) error) error {
 	if c.Scheme != "http" && c.Scheme != "https" {
 		return fmt.Errorf("`cluster.scheme` must be `http` or `https`, got %q instead for %q", c.Scheme, c.Name)
 	}
-	if c.HeartBeatInterval == 0 {
-		c.HeartBeatInterval = Duration(time.Second * 5)
+	if c.HeartBeatInterval != 0 && c.HeartBeat.Interval != defaultHeartBeat.Interval {
+		return fmt.Errorf("cannot be use `heartbeat_interval` with `heartbeat.interval`")
+	}
+	if c.HeartBeat.Interval == 0 && c.HeartBeat.Timeout == 0 && c.HeartBeat.Response == "" {
+		return fmt.Errorf("`cluster.heartbeat` cannot be unset for %q", c.Name)
 	}
 	return checkOverflow(c.XXX, fmt.Sprintf("cluster %q", c.Name))
 }
@@ -386,6 +400,41 @@ func (u *KillQueryUser) UnmarshalYAML(unmarshal func(interface{}) error) error {
 		return fmt.Errorf("`cluster.kill_query_user.name` must be specified")
 	}
 	return checkOverflow(u.XXX, "kill_query_user")
+}
+
+type HeartBeat struct {
+    // Interval is an interval of checking
+    // all cluster nodes for availability
+    // if omitted or zero - interval will be set to 5s
+    Interval Duration `yaml:"interval,omitempty"`
+
+    // Timeout is a timeout of wait response from cluster nodes
+    // if omitted or zero - interval will be set to 3s
+    Timeout Duration `yaml:"timeout,omitempty"`
+
+    // Request is a query
+    // default value is `/ping`
+    Request string `yaml:"request,omitempty"`
+
+    // Reference response from clickhouse on health check request
+    // default value is `Ok.\n`
+    Response string `yaml:"response,omitempty"`
+
+    // ToUser is the name of cluster_user
+    // whom credentials will be used for heart beat request to clickhouse
+    ToUser string `yaml:"to_user,omitempty"`
+
+    // Catches all undefined fields
+    XXX map[string]interface{} `yaml:",inline"`
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface.
+func (h *HeartBeat) UnmarshalYAML(unmarshal func(interface{}) error) error {
+    type plain HeartBeat
+    if err := unmarshal((*plain)(h)); err != nil {
+        return err
+    }
+    return checkOverflow(h.XXX, "heartbeat")
 }
 
 // User describes list of allowed users

--- a/config/config_test.go
+++ b/config/config_test.go
@@ -80,7 +80,13 @@ func TestLoadConfig(t *testing.T) {
 								MaxExecutionTime:     Duration(time.Minute),
 							},
 						},
-						HeartBeatInterval: Duration(time.Minute),
+						HeartBeat: HeartBeat{
+							Interval: Duration(time.Minute),
+							Timeout: Duration(10 * time.Second),
+							Request: "/?query=SELECT%201",
+							Response: "1\n",
+							ToUser: "web",
+						},
 					},
 					{
 						Name:   "second cluster",
@@ -111,7 +117,12 @@ func TestLoadConfig(t *testing.T) {
 								MaxQueueTime:         Duration(70 * time.Second),
 							},
 						},
-						HeartBeatInterval: Duration(5 * time.Second),
+						HeartBeat: HeartBeat{
+							Interval: Duration(5 * time.Second),
+							Timeout: Duration(3 * time.Second),
+							Request: "/ping",
+							Response: "Ok.\n",
+						},
 					},
 				},
 
@@ -223,7 +234,12 @@ func TestLoadConfig(t *testing.T) {
 								Name: "default",
 							},
 						},
-						HeartBeatInterval: Duration(5 * time.Second),
+						HeartBeat: HeartBeat{
+							Interval: Duration(5 * time.Second),
+							Timeout: Duration(3 * time.Second),
+							Request: "/ping",
+							Response: "Ok.\n",
+						},
 					},
 				},
 				Users: []User{
@@ -388,6 +404,11 @@ func TestBadConfig(t *testing.T) {
 			"empty param group params",
 			"testdata/bad.param_groups.params.yml",
 			"`param_group.params` must contain at least one param",
+		},
+		{
+			"duplicate heartbeat interval",
+			"testdata/bad.heartbeat_interval.duplicate.yml",
+			"cannot be use `heartbeat_interval` with `heartbeat.interval`",
 		},
 	}
 

--- a/config/config_test.go
+++ b/config/config_test.go
@@ -80,11 +80,12 @@ func TestLoadConfig(t *testing.T) {
 								MaxExecutionTime:     Duration(time.Minute),
 							},
 						},
+						HeartBeatInterval: Duration(time.Minute),
 						HeartBeat: HeartBeat{
 							Interval: Duration(time.Minute),
-							Timeout: Duration(10 * time.Second),
-							Request: "/ping",
-							Response: "Ok.\n",
+							Timeout:  Duration(3 * time.Second),
+							Request:  "/?query=SELECT%201",
+							Response: "1\n",
 						},
 					},
 					{
@@ -118,9 +119,25 @@ func TestLoadConfig(t *testing.T) {
 						},
 						HeartBeat: HeartBeat{
 							Interval: Duration(5 * time.Second),
-							Timeout: Duration(3 * time.Second),
-							Request: "/?query=SELECT%201",
+							Timeout:  Duration(3 * time.Second),
+							Request:  "/?query=SELECT%201",
 							Response: "1\n",
+						},
+					},
+					{
+						Name:   "thrid cluster",
+						Scheme: "http",
+						Nodes:  []string{"thrid1:8123", "thrid2:8123"},
+						ClusterUsers: []ClusterUser{
+							{
+								Name: "default",
+							},
+						},
+						HeartBeat: HeartBeat{
+							Interval: Duration(2 * time.Minute),
+							Timeout:  Duration(10 * time.Second),
+							Request:  "/ping",
+							Response: "Ok.\n",
 						},
 					},
 				},
@@ -235,8 +252,8 @@ func TestLoadConfig(t *testing.T) {
 						},
 						HeartBeat: HeartBeat{
 							Interval: Duration(5 * time.Second),
-							Timeout: Duration(3 * time.Second),
-							Request: "/?query=SELECT%201",
+							Timeout:  Duration(3 * time.Second),
+							Request:  "/?query=SELECT%201",
 							Response: "1\n",
 						},
 					},
@@ -407,7 +424,17 @@ func TestBadConfig(t *testing.T) {
 		{
 			"duplicate heartbeat interval",
 			"testdata/bad.heartbeat_interval.duplicate.yml",
-			"cannot be use `heartbeat_interval` with `heartbeat.interval`",
+			"cannot be use `heartbeat_interval` with `heartbeat` section",
+		},
+		{
+			"empty heartbeat section",
+			"testdata/bad.heartbeat_section.empty1.yml",
+			"`cluster.heartbeat` cannot be unset for \"cluster\"",
+		},
+		{
+			"empty heartbeat section with heartbeat_interval",
+			"testdata/bad.heartbeat_section.empty2.yml",
+			"cannot be use `heartbeat_interval` with `heartbeat` section",
 		},
 	}
 

--- a/config/config_test.go
+++ b/config/config_test.go
@@ -83,9 +83,8 @@ func TestLoadConfig(t *testing.T) {
 						HeartBeat: HeartBeat{
 							Interval: Duration(time.Minute),
 							Timeout: Duration(10 * time.Second),
-							Request: "/?query=SELECT%201",
-							Response: "1\n",
-							ToUser: "web",
+							Request: "/ping",
+							Response: "Ok.\n",
 						},
 					},
 					{
@@ -120,8 +119,8 @@ func TestLoadConfig(t *testing.T) {
 						HeartBeat: HeartBeat{
 							Interval: Duration(5 * time.Second),
 							Timeout: Duration(3 * time.Second),
-							Request: "/ping",
-							Response: "Ok.\n",
+							Request: "/?query=SELECT%201",
+							Response: "1\n",
 						},
 					},
 				},
@@ -237,8 +236,8 @@ func TestLoadConfig(t *testing.T) {
 						HeartBeat: HeartBeat{
 							Interval: Duration(5 * time.Second),
 							Timeout: Duration(3 * time.Second),
-							Request: "/ping",
-							Response: "Ok.\n",
+							Request: "/?query=SELECT%201",
+							Response: "1\n",
 						},
 					},
 				},

--- a/config/testdata/bad.heartbeat_interval.duplicate.yml
+++ b/config/testdata/bad.heartbeat_interval.duplicate.yml
@@ -1,0 +1,15 @@
+server:
+  http:
+    listen_addr: ":8080"
+
+users:
+  - name: "default"
+    to_cluster: "cluster"
+    to_user: "default"
+
+clusters:
+  - name: "cluster"
+    nodes: ["127.0.0.1:8123"]
+    heartbeat_interval: 1m
+    heartbeat:
+      interval: 30s

--- a/config/testdata/bad.heartbeat_section.empty1.yml
+++ b/config/testdata/bad.heartbeat_section.empty1.yml
@@ -1,0 +1,13 @@
+server:
+  http:
+    listen_addr: ":8080"
+
+users:
+  - name: "default"
+    to_cluster: "cluster"
+    to_user: "default"
+
+clusters:
+  - name: "cluster"
+    nodes: ["127.0.0.1:8123"]
+    heartbeat:

--- a/config/testdata/bad.heartbeat_section.empty2.yml
+++ b/config/testdata/bad.heartbeat_section.empty2.yml
@@ -1,0 +1,14 @@
+server:
+  http:
+    listen_addr: ":8080"
+
+users:
+  - name: "default"
+    to_cluster: "cluster"
+    to_user: "default"
+
+clusters:
+  - name: "cluster"
+    nodes: ["127.0.0.1:8123"]
+    heartbeat_interval: 1m
+    heartbeat:

--- a/config/testdata/full.yml
+++ b/config/testdata/full.yml
@@ -220,9 +220,31 @@ clusters:
     # Requests are evenly distributed among them.
     nodes: ["127.0.0.1:8123", "shard2:8123"]
 
-    # Each cluster node is checked for availability using this interval.
+    # DEPRECATED: Each cluster node is checked for availability using this interval.
     # By default each node is checked for every 5 seconds.
-    heartbeat_interval: 1m
+    # Use `heartbeat.interval`.
+    # heartbeat_interval: 1m
+
+    # User configuration for heart beat requests.
+    heartbeat:
+      # An interval for checking all cluster nodes for availability
+      # By default each node is checked for every 5 seconds.
+      interval: 1m
+
+      # A timeout of wait response from cluster nodes
+      # By default 3s
+      timeout: 10s
+
+      # The parameter to set the URI to request in a health check
+      # By default "/ping"
+      request: "/?query=SELECT%201"
+
+      # Reference response from clickhouse on health check request
+      # By default "Ok.\n"
+      response: "1\n"
+
+      # The name of cluster_user whom credentials will be used for heart beat request to clickhouse.
+      to_user: "web"
 
     # Timed out queries are killed using this user.
     # By default `default` user is used.

--- a/config/testdata/full.yml
+++ b/config/testdata/full.yml
@@ -226,6 +226,7 @@ clusters:
     # heartbeat_interval: 1m
 
     # User configuration for heart beat requests.
+    # Credentials of the first user in clusters.users will be used for heart beat requests to clickhouse.
     heartbeat:
       # An interval for checking all cluster nodes for availability
       # By default each node is checked for every 5 seconds.
@@ -236,15 +237,12 @@ clusters:
       timeout: 10s
 
       # The parameter to set the URI to request in a health check
-      # By default "/ping"
-      request: "/?query=SELECT%201"
+      # By default "/?query=SELECT%201"
+      request: "/ping"
 
       # Reference response from clickhouse on health check request
-      # By default "Ok.\n"
-      response: "1\n"
-
-      # The name of cluster_user whom credentials will be used for heart beat request to clickhouse.
-      to_user: "web"
+      # By default "1\n"
+      response: "Ok.\n"
 
     # Timed out queries are killed using this user.
     # By default `default` user is used.

--- a/config/testdata/full.yml
+++ b/config/testdata/full.yml
@@ -223,26 +223,7 @@ clusters:
     # DEPRECATED: Each cluster node is checked for availability using this interval.
     # By default each node is checked for every 5 seconds.
     # Use `heartbeat.interval`.
-    # heartbeat_interval: 1m
-
-    # User configuration for heart beat requests.
-    # Credentials of the first user in clusters.users will be used for heart beat requests to clickhouse.
-    heartbeat:
-      # An interval for checking all cluster nodes for availability
-      # By default each node is checked for every 5 seconds.
-      interval: 1m
-
-      # A timeout of wait response from cluster nodes
-      # By default 3s
-      timeout: 10s
-
-      # The parameter to set the URI to request in a health check
-      # By default "/?query=SELECT%201"
-      request: "/ping"
-
-      # Reference response from clickhouse on health check request
-      # By default "1\n"
-      response: "Ok.\n"
+    heartbeat_interval: 1m
 
     # Timed out queries are killed using this user.
     # By default `default` user is used.
@@ -282,3 +263,25 @@ clusters:
         max_queue_size: 50
         max_queue_time: 70s
         allowed_networks: ["office"]
+
+  - name: "thrid cluster"
+    nodes: ["thrid1:8123", "thrid2:8123"]
+
+    # User configuration for heart beat requests.
+    # Credentials of the first user in clusters.users will be used for heart beat requests to clickhouse.
+    heartbeat:
+      # An interval for checking all cluster nodes for availability
+      # By default each node is checked for every 5 seconds.
+      interval: 2m
+
+      # A timeout of wait response from cluster nodes
+      # By default 3s
+      timeout: 10s
+
+      # The parameter to set the URI to request in a health check
+      # By default "/?query=SELECT%201"
+      request: "/ping"
+
+      # Reference response from clickhouse on health check request
+      # By default "1\n"
+      response: "Ok.\n"

--- a/heartbeat.go
+++ b/heartbeat.go
@@ -1,0 +1,85 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/Vertamedia/chproxy/config"
+)
+
+type heartBeat struct {
+	interval time.Duration
+	timeout time.Duration
+	request string
+	response string
+	user string
+	password string
+}
+
+func newHeartBeat(c config.HeartBeat, clusterUsers map[string]*clusterUser, oldInterval time.Duration) (*heartBeat, error) {
+	heartBeatPassword, err := func() (string, error) {
+		if len(c.ToUser) == 0 {
+			return "", nil
+		}
+		hu, ok := clusterUsers[c.ToUser]
+		if !ok {
+			return "", fmt.Errorf("unknown `to_user` %q", c.ToUser)
+		}
+		return hu.password, nil
+	}()
+	if err != nil {
+		return nil, err
+	}
+
+	newHB := &heartBeat{
+		interval:	time.Duration(c.Interval),
+		timeout:	time.Duration(c.Timeout),
+		request:	c.Request,
+		response:	c.Response,
+		user:       c.ToUser,
+		password:   heartBeatPassword,
+	}
+
+	if oldInterval != 0 {
+		newHB.interval = oldInterval
+	}
+
+	return newHB, nil
+}
+
+func (hb *heartBeat) isHealthy(addr string) error {
+	req, err := http.NewRequest("GET", addr + hb.request, nil)
+	if err != nil {
+		return err
+	}
+
+	if hb.user != "" {
+		req.SetBasicAuth(hb.user, hb.password)
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), hb.timeout)
+	defer cancel()
+	req = req.WithContext(ctx)
+
+	startTime := time.Now()
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("cannot send request in %s: %s", time.Since(startTime), err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("non-200 status code: %s", resp.Status)
+	}
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("cannot read response in %s: %s", time.Since(startTime), err)
+	}
+	r := string(body)
+	if r != hb.response {
+		return fmt.Errorf("unexpected response: %s", r)
+	}
+	return nil
+}

--- a/heartbeat.go
+++ b/heartbeat.go
@@ -19,28 +19,14 @@ type heartBeat struct {
 	password string
 }
 
-func newHeartBeat(c config.HeartBeat, clusterUsers map[string]*clusterUser, oldInterval time.Duration) (*heartBeat, error) {
-	heartBeatPassword, err := func() (string, error) {
-		if len(c.ToUser) == 0 {
-			return "", nil
-		}
-		hu, ok := clusterUsers[c.ToUser]
-		if !ok {
-			return "", fmt.Errorf("unknown `to_user` %q", c.ToUser)
-		}
-		return hu.password, nil
-	}()
-	if err != nil {
-		return nil, err
-	}
-
+func newHeartBeat(c config.HeartBeat, firstClusterUser config.ClusterUser, oldInterval time.Duration) (*heartBeat, error) {
 	newHB := &heartBeat{
 		interval:	time.Duration(c.Interval),
 		timeout:	time.Duration(c.Timeout),
 		request:	c.Request,
 		response:	c.Response,
-		user:       c.ToUser,
-		password:   heartBeatPassword,
+		user:		firstClusterUser.Name,
+		password:	firstClusterUser.Password,
 	}
 
 	if oldInterval != 0 {

--- a/heartbeat.go
+++ b/heartbeat.go
@@ -12,36 +12,30 @@ import (
 
 type heartBeat struct {
 	interval time.Duration
-	timeout time.Duration
-	request string
+	timeout  time.Duration
+	request  string
 	response string
-	user string
+	user     string
 	password string
 }
 
-func newHeartBeat(c config.HeartBeat, firstClusterUser config.ClusterUser, oldInterval time.Duration) (*heartBeat, error) {
+func newHeartBeat(c config.HeartBeat, firstClusterUser config.ClusterUser) *heartBeat {
 	newHB := &heartBeat{
-		interval:	time.Duration(c.Interval),
-		timeout:	time.Duration(c.Timeout),
-		request:	c.Request,
-		response:	c.Response,
-		user:		firstClusterUser.Name,
-		password:	firstClusterUser.Password,
+		interval: time.Duration(c.Interval),
+		timeout:  time.Duration(c.Timeout),
+		request:  c.Request,
+		response: c.Response,
+		user:     firstClusterUser.Name,
+		password: firstClusterUser.Password,
 	}
-
-	if oldInterval != 0 {
-		newHB.interval = oldInterval
-	}
-
-	return newHB, nil
+	return newHB
 }
 
 func (hb *heartBeat) isHealthy(addr string) error {
-	req, err := http.NewRequest("GET", addr + hb.request, nil)
+	req, err := http.NewRequest("GET", addr+hb.request, nil)
 	if err != nil {
 		return err
 	}
-
 	if hb.user != "" {
 		req.SetBasicAuth(hb.user, hb.password)
 	}

--- a/heartbeat_test.go
+++ b/heartbeat_test.go
@@ -1,0 +1,126 @@
+package main
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"github.com/Vertamedia/chproxy/config"
+)
+
+
+var (
+	hbHandler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Path == "/ping" {
+			fmt.Fprintln(w, "Ok.")
+			return
+		}
+		qid := r.URL.Query().Get("query")
+		if len(qid) != 0 {
+			fmt.Fprintln(w, "1")
+			return
+		}
+		fmt.Fprintln(w, "wrong")
+	})
+
+	fakeHBServer = httptest.NewServer(hbHandler)
+
+	clusterCfg = config.Cluster{
+		Name:   "cluster",
+		Scheme: "http",
+		Replicas: []config.Replica{
+			{
+				Nodes: []string{fakeHBServer.URL},
+			},
+		},
+		ClusterUsers: []config.ClusterUser{
+			{
+				Name: "web",
+				Password: "123",
+			},
+		},
+		HeartBeatInterval: config.Duration(10 * time.Second),
+		HeartBeat: config.HeartBeat{
+			Timeout: config.Duration(3 * time.Second),
+			Request: "/ping",
+			Response: "Ok.\n",
+		},
+	}
+
+	heartBeatFullCfg = config.HeartBeat{
+		Interval: config.Duration(20 * time.Second),
+		Timeout: config.Duration(30 * time.Second),
+		Request: "/?query=SELECT%201",
+		Response: "1\n",
+		ToUser: "web",
+	}
+
+	heartBeatWrongUserCfg = config.HeartBeat{
+		ToUser: "test",
+	}
+
+	heartBeatWrongResponseCfg = config.HeartBeat{
+		Interval: config.Duration(20 * time.Second),
+		Timeout: config.Duration(30 * time.Second),
+		Request: "/wrongQuery",
+		Response: "Ok.\n",
+	}
+)
+
+func TestNewHeartBeat(t *testing.T) {
+	c, err := newCluster(clusterCfg)
+	if err != nil {
+		t.Fatalf("error while initialize claster: %s", err)
+	}
+	testCompareNum(t, "heartbeat_interval", int64(c.heartBeat.interval/time.Microsecond), int64(time.Duration(10 * time.Second)/time.Microsecond))
+
+	hb, err := newHeartBeat(heartBeatFullCfg, c.users, 0)
+	if err != nil {
+		t.Fatalf("error while initialize heartbeat: %s", err)
+	}
+	testCompareNum(t, "heartbeat.interval", int64(hb.interval/time.Microsecond), int64(time.Duration(20 * time.Second)/time.Microsecond))
+	testCompareNum(t, "heartbeat.timeout", int64(hb.timeout/time.Microsecond), int64(time.Duration(30 * time.Second)/time.Microsecond))
+	testCompareStr(t, "heartbeat.request", hb.request, "/?query=SELECT%201")
+	testCompareStr(t, "heartbeat.response", hb.response, "1\n")
+	testCompareStr(t, "heartbeat.user", hb.user, "web")
+	testCompareStr(t, "heartbeat.password", hb.password, "123")
+	if check := c.heartBeat.isHealthy(fakeHBServer.URL); check != nil {
+		t.Fatalf("ping request error `%q`", check)
+	}
+	if check := hb.isHealthy(fakeHBServer.URL); check != nil {
+		t.Fatalf("query request error `%q`", check)
+	}
+
+	_, err = newHeartBeat(heartBeatWrongUserCfg, c.users, 0)
+	if err == nil {
+		t.Fatalf("error expected")
+	}
+	testCompareStr(t, "error", err.Error(), "unknown `to_user` \"test\"")
+
+	hbWrong, err := newHeartBeat(heartBeatWrongResponseCfg, c.users, 0)
+	if err != nil {
+		t.Fatalf("error while initialize heartbeat: %s", err)
+	}
+	check := hbWrong.isHealthy(fakeHBServer.URL)
+	if check == nil {
+		t.Fatalf("heartbeat error expected")
+	}
+	testCompareStr(t, "heartbeat error", check.Error(), "unexpected response: wrong\n")
+}
+
+func testCompareNum(t *testing.T, name string, value int64, expected int64) {
+	t.Helper()
+	if value != expected {
+		t.Fatalf("expected %s for host: `%d`; got: `%d`", name, expected, value)
+	}
+}
+
+func testCompareStr(t *testing.T, name string, value string, expected string) {
+	t.Helper()
+	if value != expected {
+		t.Fatalf("expected %s for host: `%q`; got: `%q`", name, expected, value)
+	}
+}
+

--- a/heartbeat_test.go
+++ b/heartbeat_test.go
@@ -10,7 +10,6 @@ import (
 	"github.com/Vertamedia/chproxy/config"
 )
 
-
 var (
 	hbHandler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		if r.URL.Path == "/ping" {
@@ -37,29 +36,29 @@ var (
 		},
 		ClusterUsers: []config.ClusterUser{
 			{
-				Name: "web",
+				Name:     "web",
 				Password: "123",
 			},
 		},
-		HeartBeatInterval: config.Duration(10 * time.Second),
 		HeartBeat: config.HeartBeat{
-			Timeout: config.Duration(3 * time.Second),
-			Request: "/ping",
+			Interval: config.Duration(5 * time.Second),
+			Timeout:  config.Duration(3 * time.Second),
+			Request:  "/ping",
 			Response: "Ok.\n",
 		},
 	}
 
 	heartBeatFullCfg = config.HeartBeat{
 		Interval: config.Duration(20 * time.Second),
-		Timeout: config.Duration(30 * time.Second),
-		Request: "/?query=SELECT%201",
+		Timeout:  config.Duration(30 * time.Second),
+		Request:  "/?query=SELECT%201",
 		Response: "1\n",
 	}
 
 	heartBeatWrongResponseCfg = config.HeartBeat{
 		Interval: config.Duration(20 * time.Second),
-		Timeout: config.Duration(30 * time.Second),
-		Request: "/wrongQuery",
+		Timeout:  config.Duration(30 * time.Second),
+		Request:  "/wrongQuery",
 		Response: "Ok.\n",
 	}
 )
@@ -69,14 +68,11 @@ func TestNewHeartBeat(t *testing.T) {
 	if err != nil {
 		t.Fatalf("error while initialize claster: %s", err)
 	}
-	testCompareNum(t, "heartbeat_interval", int64(c.heartBeat.interval/time.Microsecond), int64(time.Duration(10 * time.Second)/time.Microsecond))
+	testCompareNum(t, "cluster.heartbeat.interval", int64(c.heartBeat.interval/time.Microsecond), int64(time.Duration(5*time.Second)/time.Microsecond))
 
-	hb, err := newHeartBeat(heartBeatFullCfg, clusterCfg.ClusterUsers[0], 0)
-	if err != nil {
-		t.Fatalf("error while initialize heartbeat: %s", err)
-	}
-	testCompareNum(t, "heartbeat.interval", int64(hb.interval/time.Microsecond), int64(time.Duration(20 * time.Second)/time.Microsecond))
-	testCompareNum(t, "heartbeat.timeout", int64(hb.timeout/time.Microsecond), int64(time.Duration(30 * time.Second)/time.Microsecond))
+	hb := newHeartBeat(heartBeatFullCfg, clusterCfg.ClusterUsers[0])
+	testCompareNum(t, "heartbeat.interval", int64(hb.interval/time.Microsecond), int64(time.Duration(20*time.Second)/time.Microsecond))
+	testCompareNum(t, "heartbeat.timeout", int64(hb.timeout/time.Microsecond), int64(time.Duration(30*time.Second)/time.Microsecond))
 	testCompareStr(t, "heartbeat.request", hb.request, "/?query=SELECT%201")
 	testCompareStr(t, "heartbeat.response", hb.response, "1\n")
 	testCompareStr(t, "heartbeat.user", hb.user, "web")
@@ -88,10 +84,7 @@ func TestNewHeartBeat(t *testing.T) {
 		t.Fatalf("query request error `%q`", check)
 	}
 
-	hbWrong, err := newHeartBeat(heartBeatWrongResponseCfg, clusterCfg.ClusterUsers[0], 0)
-	if err != nil {
-		t.Fatalf("error while initialize heartbeat: %s", err)
-	}
+	hbWrong := newHeartBeat(heartBeatWrongResponseCfg, clusterCfg.ClusterUsers[0])
 	check := hbWrong.isHealthy(fakeHBServer.URL)
 	if check == nil {
 		t.Fatalf("heartbeat error expected")
@@ -112,4 +105,3 @@ func testCompareStr(t *testing.T, name string, value string, expected string) {
 		t.Fatalf("expected %s for host: `%q`; got: `%q`", name, expected, value)
 	}
 }
-

--- a/heartbeat_test.go
+++ b/heartbeat_test.go
@@ -54,11 +54,6 @@ var (
 		Timeout: config.Duration(30 * time.Second),
 		Request: "/?query=SELECT%201",
 		Response: "1\n",
-		ToUser: "web",
-	}
-
-	heartBeatWrongUserCfg = config.HeartBeat{
-		ToUser: "test",
 	}
 
 	heartBeatWrongResponseCfg = config.HeartBeat{
@@ -76,7 +71,7 @@ func TestNewHeartBeat(t *testing.T) {
 	}
 	testCompareNum(t, "heartbeat_interval", int64(c.heartBeat.interval/time.Microsecond), int64(time.Duration(10 * time.Second)/time.Microsecond))
 
-	hb, err := newHeartBeat(heartBeatFullCfg, c.users, 0)
+	hb, err := newHeartBeat(heartBeatFullCfg, clusterCfg.ClusterUsers[0], 0)
 	if err != nil {
 		t.Fatalf("error while initialize heartbeat: %s", err)
 	}
@@ -93,13 +88,7 @@ func TestNewHeartBeat(t *testing.T) {
 		t.Fatalf("query request error `%q`", check)
 	}
 
-	_, err = newHeartBeat(heartBeatWrongUserCfg, c.users, 0)
-	if err == nil {
-		t.Fatalf("error expected")
-	}
-	testCompareStr(t, "error", err.Error(), "unknown `to_user` \"test\"")
-
-	hbWrong, err := newHeartBeat(heartBeatWrongResponseCfg, c.users, 0)
+	hbWrong, err := newHeartBeat(heartBeatWrongResponseCfg, clusterCfg.ClusterUsers[0], 0)
 	if err != nil {
 		t.Fatalf("error while initialize heartbeat: %s", err)
 	}

--- a/proxy_test.go
+++ b/proxy_test.go
@@ -22,7 +22,7 @@ import (
 )
 
 const (
-	okResponse = "Ok.\n"
+	okResponse = "1\n"
 )
 
 var goodCfg = &config.Config{

--- a/proxy_test.go
+++ b/proxy_test.go
@@ -21,6 +21,10 @@ import (
 	"github.com/Vertamedia/chproxy/config"
 )
 
+const (
+	okResponse = "Ok.\n"
+)
+
 var goodCfg = &config.Config{
 	Clusters: []config.Cluster{
 		{

--- a/scope.go
+++ b/scope.go
@@ -630,8 +630,9 @@ func (h *host) runHeartbeat(done <-chan struct{}) {
 		"replica":      h.replica.name,
 		"cluster_node": h.addr.Host,
 	}
+	hb := h.replica.cluster.heartBeat
 	heartbeat := func() {
-		if err := isHealthy(h.addr.String()); err == nil {
+		if err := hb.isHealthy(h.addr.String()); err == nil {
 			atomic.StoreUint32(&h.active, uint32(1))
 			hostHealth.With(label).Set(1)
 		} else {
@@ -641,12 +642,11 @@ func (h *host) runHeartbeat(done <-chan struct{}) {
 		}
 	}
 	heartbeat()
-	interval := h.replica.cluster.heartBeatInterval
 	for {
 		select {
 		case <-done:
 			return
-		case <-time.After(interval):
+		case <-time.After(hb.interval):
 			heartbeat()
 		}
 	}
@@ -714,7 +714,7 @@ type cluster struct {
 	killQueryUserName     string
 	killQueryUserPassword string
 
-	heartBeatInterval time.Duration
+	heartBeat *heartBeat
 }
 
 func newCluster(c config.Cluster) (*cluster, error) {
@@ -726,12 +726,17 @@ func newCluster(c config.Cluster) (*cluster, error) {
 		clusterUsers[cu.Name] = newClusterUser(cu)
 	}
 
+	heartBeat, err := newHeartBeat(c.HeartBeat, clusterUsers, time.Duration(c.HeartBeatInterval))
+	if err != nil {
+		return nil, fmt.Errorf("cannot initialize heartbeat: %s", err)
+	}
+
 	newC := &cluster{
 		name:                  c.Name,
 		users:                 clusterUsers,
 		killQueryUserName:     c.KillQueryUser.Name,
 		killQueryUserPassword: c.KillQueryUser.Password,
-		heartBeatInterval:     time.Duration(c.HeartBeatInterval),
+		heartBeat:             heartBeat,
 	}
 
 	replicas, err := newReplicas(c.Replicas, c.Nodes, c.Scheme, newC)

--- a/scope.go
+++ b/scope.go
@@ -726,13 +726,7 @@ func newCluster(c config.Cluster) (*cluster, error) {
 		clusterUsers[cu.Name] = newClusterUser(cu)
 	}
 
-	if (len(c.ClusterUsers) == 0) {
-		return nil, fmt.Errorf("cannot initialize heartbeat: clusters configuration must contain at least 1 user")
-	}
-	heartBeat, err := newHeartBeat(c.HeartBeat, c.ClusterUsers[0], time.Duration(c.HeartBeatInterval))
-	if err != nil {
-		return nil, fmt.Errorf("cannot initialize heartbeat: %s", err)
-	}
+	heartBeat := newHeartBeat(c.HeartBeat, c.ClusterUsers[0])
 
 	newC := &cluster{
 		name:                  c.Name,

--- a/scope.go
+++ b/scope.go
@@ -726,7 +726,10 @@ func newCluster(c config.Cluster) (*cluster, error) {
 		clusterUsers[cu.Name] = newClusterUser(cu)
 	}
 
-	heartBeat, err := newHeartBeat(c.HeartBeat, clusterUsers, time.Duration(c.HeartBeatInterval))
+	if (len(c.ClusterUsers) == 0) {
+		return nil, fmt.Errorf("cannot initialize heartbeat: clusters configuration must contain at least 1 user")
+	}
+	heartBeat, err := newHeartBeat(c.HeartBeat, c.ClusterUsers[0], time.Duration(c.HeartBeatInterval))
 	if err != nil {
 		return nil, fmt.Errorf("cannot initialize heartbeat: %s", err)
 	}

--- a/utils.go
+++ b/utils.go
@@ -3,14 +3,12 @@ package main
 import (
 	"bytes"
 	"compress/gzip"
-	"context"
 	"fmt"
 	"io"
 	"io/ioutil"
 	"net/http"
 	"sort"
 	"strings"
-	"time"
 
 	"github.com/Vertamedia/chproxy/chdecompressor"
 	"github.com/Vertamedia/chproxy/log"
@@ -36,41 +34,6 @@ func getAuth(req *http.Request) (string, string) {
 	}
 	// if still no credentials - treat it as `default` user request
 	return "default", ""
-}
-
-const (
-	okResponse       = "Ok.\n"
-	isHealthyTimeout = 3 * time.Second
-)
-
-func isHealthy(addr string) error {
-	req, err := http.NewRequest("GET", addr, nil)
-	if err != nil {
-		return err
-	}
-	ctx, cancel := context.WithTimeout(context.Background(), isHealthyTimeout)
-	defer cancel()
-	req = req.WithContext(ctx)
-
-	startTime := time.Now()
-	resp, err := http.DefaultClient.Do(req)
-	if err != nil {
-		return fmt.Errorf("cannot send request in %s: %s", time.Since(startTime), err)
-	}
-	defer resp.Body.Close()
-
-	if resp.StatusCode != http.StatusOK {
-		return fmt.Errorf("non-200 status code: %s", resp.Status)
-	}
-	body, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		return fmt.Errorf("cannot read response in %s: %s", time.Since(startTime), err)
-	}
-	r := string(body)
-	if r != okResponse {
-		return fmt.Errorf("unexpected response: %s", r)
-	}
-	return nil
 }
 
 // getQuerySnippet returns query snippet.

